ui/perfherder/graphs/GraphsView.jsx

```diff
@@ -0,0 +1,448 @@
+  }
+
+  // TODO should add a custom time range option based on query param
+  getDefaultTimeRange = () => {
+    const { $stateParams } = this.props;
+
+    const defaultValue = $stateParams.timerange
+      ? parseInt($stateParams.timerange, 10)
+      : phDefaultTimeRangeValue;
+    return phTimeRanges.find(time => time.value === defaultValue);
+  };
+
+  async getData() {
+    const [projects, frameworks] = await Promise.all([
+      getData(getApiUrl(repoEndpoint)),
+      getData(getApiUrl(endpoints.frameworks)),
+    ]);
+
+    const updates = {
+      ...processResponse(projects, 'projects'),
+      ...processResponse(frameworks, 'frameworks'),
+    };
+    this.setState(updates);
+  }
+
+  checkQueryParams = () => {
+    const {
+      series,
+      zoom,
+      selected,
+      highlightAlerts,
+      highlightedRevisions,
+    } = this.props.$stateParams;
+
+    const updates = {};
+
+    if (series) {
+      const _series = typeof series === 'string' ? [series] : series;
+      const seriesParams = this.parseSeriesParam(_series);
+      this.getTestData(seriesParams);
+    }
+
+    if (highlightAlerts) {
+      updates.highlightAlerts = Boolean(parseInt(highlightAlerts, 10));
+    }
+
+    if (highlightedRevisions) {
+      updates.highlightedRevisions =
+        typeof highlightedRevisions === 'string'
+          ? [highlightedRevisions]
+          : highlightedRevisions;
+    }
+
+    if (zoom) {
+      const zoomArray = zoom.replace(/[[{}\]"]+/g, '').split(',');
+      const zoomObject = {
+        x: zoomArray.slice(0, 2),
+        y: zoomArray.slice(2, 4),
+      };
+      updates.zoom = zoomObject;
+    }
+
+    if (selected) {
+      const tooltipArray = selected.replace(/[[]"]/g, '').split(',');
+      // TODO keys should reflect perf/sumary/ data
+      const tooltipValues = {
+        projectName: tooltipArray[0],
+        signatureId: parseInt(tooltipArray[1], 10),
+        resultSetId: parseInt(tooltipArray[2], 10),
+        id: parseInt(tooltipArray[3], 10),
+        frameworkId: parseInt(tooltipArray[4], 10) || 1,
+      };
+      updates.selectedDataPoint = tooltipValues;
+    }
+
+    this.setState(updates);
+  };
+
+  createSeriesParams = series => {
+    const { repository_name, signature_id, framework_id } = series;
+    const { timeRange } = this.state;
+
+    return {
+      repository: repository_name,
+      signature: signature_id,
+      framework: framework_id,
+      interval: timeRange.value,
+      all_data: true,
+    };
+  };
+
+  getTestData = async (newDisplayedTests = []) => {
+    const { testData } = this.state;
+    const tests = newDisplayedTests.length ? newDisplayedTests : testData;
+    this.setState({ loading: true });
+
+    const responses = await Promise.all(
+      tests.map(series =>
+        getData(
+          createApiUrl(perfSummaryEndpoint, this.createSeriesParams(series)),
+        ),
+      ),
+    );
+    const errorMessages = processErrors(responses);
```

Comment:
I think this is the pattern you referenced in your comments on my other PR.  I'll adjust to using this pattern.  It looks tidy.  :)

Reply:
Well there's two - `processResponse` which is for more general use. This `processErrors` is for a specific use case of dealing with only errors in the array returned when using `Promise.all` (and I either don't care about the response if the update was OK or I need to munge the data before setting state).